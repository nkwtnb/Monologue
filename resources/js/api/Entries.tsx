import axios from "axios";
import { Entry } from "@interface/Entry";

interface Props {
  name: string | undefined;
  filter: "post" | "like" | "media";
}

/**
 * パラメータに応じて、全投稿、いいねした投稿などを取得する
 */
export const getEntries = async ({ name, filter }: Props): Promise<Entry[]> => {
  // ホーム画面時
  if (name === "") {
    const entries = (await axios.get(`/api/words`)).data as Entry[];
    return entries;
    // ユーザー別 > いいね
  } else if (filter === "like") {
    const likeEntries = (await axios.get(`/api/words/user/${name}/likes`)).data as Entry[];
    return likeEntries;
    // ユーザー別 > 投稿 or メディア
  } else {
    const entries = (await axios.get(`/api/words/user/${name}/posts`)).data as Entry[];
    console.log(entries);
    return entries;
  }
}

/**
 * 指定されたポストIDの投稿を取得する
 * @param postId 
 */
export const getEntry = async (postId: string | undefined): Promise<{entry: Entry, replies: Entry[]}> => {
  const resp = (await axios.get("/api/words/post/" + postId)).data;
  const entry: Entry = resp.entries[0];
  const replies: Entry[] = resp.replies;
  return {
    entry,
    replies
  };
}


// /**
//  * 指定されたポストIDの投稿を取得する
//  * @param postId 
//  */
//  export const getEntry = async (postId: string | undefined): Promise<Entry[]> => {
//   const resp = (await axios.get("/api/words/post/" + postId)).data as Entry[];
//   return resp;
// }

// /**
//  * 対象の投稿を、投稿／コメントに分ける
//  * @param _entry
//  */
// export const separateEntry = (_entry: any): {entry: Entry, replies: Entry[]} => {
//   const entry: Entry = _entry.entries[0];
//   const replies: Entry[] = _entry.replies;
//   return {
//     entry,
//     replies
//   };
// }

/**
 * じぶんの「いいね」を取得する
 * @param authName 
 * @returns 
 */
export const getLikes = async (authName: string | undefined) => {
  const likeEntries = authName === "" ? [] : (await axios.get(`/api/words/user/${authName}/likes`)).data as Entry[];
  return likeEntries;
}

/**
 * 自分が「いいね」した投稿にフラグを立てる
 * @returns 
 */
export const setLikeStatus = (entries: Entry[], likeEntries: Entry[]): Entry[] => {
  const resp = entries.map(entry => {
    for (let i = 0; i < likeEntries.length; i++) {
      const likeEntry = likeEntries[i];
      if (entry.id === likeEntry.id) {
        entry.isLike = true;
        break;
      }
    }
    return entry;
  });
  return resp;
}
