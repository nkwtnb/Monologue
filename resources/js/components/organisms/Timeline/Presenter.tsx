import React from 'react';
import Post from '../Post';
import { Entry } from "@interface/Entry";

interface Props {
  isLoading: boolean
  data: Entry[]
  error?: any
}

export default (props: Props): JSX.Element => {
  if (props.isLoading) {
    return <span data-testid="loading">読み込み中</span>;
  }
  if (props.data.length === 0) {
    return <div data-testid="no-data">対象の投稿がありません</div>;
  } else {
    return (
      <>
        {
          props.data.map((entry: Entry, index) => {
            return (
              <div data-testid="post" className='px-0 mb-1' key={index}>
                <Post
                  id={entry.id}
                  name={entry.name}
                  created_at={entry.created_at}
                  avatar={entry.avatar}
                  words={entry.words}
                  ogp_title={entry.ogp_title}
                  ogp_description={entry.ogp_description}
                  ogp_image={entry.ogp_image}
                  images={entry.images}
                  isLike={entry.isLike}
                  likes={entry.likes}
                  replyCount={entry.replyCount}
                  isDialog={false}
                />
              </div>
            )
          })
        }
      </>
    );
  }
}
