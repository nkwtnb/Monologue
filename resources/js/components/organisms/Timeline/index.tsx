import React, { useContext } from 'react';
import { AuthContext } from '../../../Context';
import useRequest from '../../../hooks/useRequest';
import Presenter from './Presenter';

interface Props {
  name: string | undefined;
  filter: "post" | "like" | "media";
}

export default (props: Props): JSX.Element => {
  const {authState, setAuthState} = useContext(AuthContext);
  const {data, error, isLoading} = useRequest({name: props.name, filter: props.filter, authState: authState});

  return <Presenter data={data} error={error} isLoading={isLoading} />
}
