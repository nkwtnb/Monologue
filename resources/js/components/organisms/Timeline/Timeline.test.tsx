import React from "react";
import Timeline from "./Presenter";
import { cleanup, render, screen } from "@testing-library/react";
import * as entryUtil from "@api/Entries";
import { BrowserRouter } from "react-router-dom";

afterEach(() => cleanup());
describe("Timeline unit test", () => {
  it("配列で受け取ったメッセージが表示される > メッセージ表示[読み込み中]", async () => {
    // レンダリング、テスト
    const result = render(
      <BrowserRouter>
        <Timeline data={[]} isLoading={true}></Timeline>
      </BrowserRouter>
    );
    expect(screen.getByTestId("loading")).toBeTruthy();
  });
  it("配列で受け取ったメッセージが表示される > メッセージ表示[投稿がありません]", async () => {
    // レンダリング、テスト
    const result = render(
      <BrowserRouter>
        <Timeline data={[]} isLoading={false}></Timeline>
      </BrowserRouter>
    );
    expect(screen.getByTestId("no-data")).toBeTruthy();
  });
  it("配列で受け取ったメッセージが表示される > データが表示される", async () => {
    const resolveValue = [
      {
        name: "username",
        id: 1,
        avatar: "",
        words: "メッセージ１",
        ogp_title: "",
        ogp_description: "",
        ogp_image: "",
        images: [],
        created_at: "",
        likes: 1,
        isLike: false,
        replyCount: 0
      }
    ]
    // データ取得処理モック化
    const mockGetEntries = jest.spyOn(entryUtil, "getEntries");
    mockGetEntries.mockResolvedValue(Promise.resolve(resolveValue));
    const mockGetLikes = jest.spyOn(entryUtil, "getLikes");
    mockGetLikes.mockResolvedValue(Promise.resolve(resolveValue));

    // レンダリング、テスト
    const result = render(
      <BrowserRouter>
        <Timeline data={resolveValue} isLoading={false}></Timeline>
      </BrowserRouter>
    );
    expect(screen.getByTestId("post")).toBeTruthy();
    expect(result.asFragment()).toMatchSnapshot();
  });
})