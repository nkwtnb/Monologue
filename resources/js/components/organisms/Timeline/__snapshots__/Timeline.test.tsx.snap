// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline unit test 配列で受け取ったメッセージが表示される > データが表示される 1`] = `
<DocumentFragment>
  <div
    class="px-0 mb-1"
    data-testid="post"
  >
    <div
      class="sc-crHmcD cximqd justify-content-center entry-card"
      id="card"
    >
      <div
        class="container-fluid"
      >
        <div
          class="row flex-nowrap"
        >
          <div
            class="sc-ksdxgE ccwIwn col-2 mt-2"
          >
            <a
              href="/user/username"
            >
              <div
                class="sc-hBUSln jbvfnu d-flex align-items-center justify-content-center"
              >
                <span
                  class="sc-bdvvtL kLovev circle-icon"
                  data-testid="circle-icon"
                >
                  <img
                    class="sc-dkPtRN jxeJUk d-flex justify-content-center align-items-center"
                    src=""
                  />
                </span>
              </div>
            </a>
          </div>
          <div
            class="col-10 flex-grow-1"
          >
            <div
              class="sc-egiyK fPWVwC"
            >
              <span>
                username
              </span>
              <span
                class="sc-fotOHu hWMOtg created-at"
              />
            </div>
            <div
              class="sc-bqiRlB edkhKt"
            >
              メッセージ１
            </div>
          </div>
        </div>
        <div
          class="row mt-2 mb-2"
        >
          <div
            class="col d-flex justify-content-center"
          >
            <div>
              <div
                class="sc-eCImPb bZEUAX"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-comment fa-w-16 "
                  data-icon="comment"
                  data-prefix="far"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 32C114.6 32 0 125.1 0 240c0 47.6 19.9 91.2 52.9 126.3C38 405.7 7 439.1 6.5 439.5c-6.6 7-8.4 17.2-4.6 26S14.4 480 24 480c61.5 0 110-25.7 139.1-46.3C192 442.8 223.2 448 256 448c141.4 0 256-93.1 256-208S397.4 32 256 32zm0 368c-26.7 0-53.1-4.1-78.4-12.1l-22.7-7.2-19.5 13.8c-14.3 10.1-33.9 21.4-57.5 29 7.3-12.1 14.4-25.7 19.9-40.2l10.6-28.1-20.6-21.8C69.7 314.1 48 282.2 48 240c0-88.2 93.3-160 208-160s208 71.8 208 160-93.3 160-208 160z"
                    fill="currentColor"
                  />
                </svg>
                <span
                  class="sc-hKwDye idVSRS"
                >
                  0
                </span>
              </div>
            </div>
          </div>
          <div
            class="col d-flex justify-content-center"
          >
            <div
              class="sc-gKclnd coqalH"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-heart fa-w-16 "
                data-icon="heart"
                data-prefix="far"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M458.4 64.3C400.6 15.7 311.3 23 256 79.3 200.7 23 111.4 15.6 53.6 64.3-21.6 127.6-10.6 230.8 43 285.5l175.4 178.7c10 10.2 23.4 15.9 37.6 15.9 14.3 0 27.6-5.6 37.6-15.8L469 285.6c53.5-54.7 64.7-157.9-10.6-221.3zm-23.6 187.5L259.4 430.5c-2.4 2.4-4.4 2.4-6.8 0L77.2 251.8c-36.5-37.2-43.9-107.6 7.3-150.7 38.9-32.7 98.9-27.8 136.5 10.5l35 35.7 35-35.7c37.8-38.5 97.8-43.2 136.5-10.6 51.1 43.1 43.5 113.9 7.3 150.8z"
                  fill="currentColor"
                />
              </svg>
              <span
                class="sc-jRQBWg hFKzTl"
              >
                1
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
